crates/bevy_ecs/src/query/iter.rs

```diff
@@ -398,3 +400,41 @@ where
         }
     }
 }
+
+/// An [`Iterator`] over the results of a query that match a list of entities.
+///
+/// This struct is created by the [`Query::many_iter`](crate::system::Query::many_iter) method.
+pub struct QueryManyIter<'w, 's, I, Q, F, T>
+where
+    T: Borrow<Entity>,
+    I: Iterator<Item = T>,
+    Q: WorldQuery,
+    F: WorldQuery,
+{
+    pub(crate) entities: I,
+    pub(crate) query: &'w Query<'w, 's, Q, F>,
```

Comment:
Potential optimization: unpack `Query::get` and cache the required parts inside the iterator.

 - Use `QueryState` over `Query`
 - initialize the fetches on construction (like QueryIter)
 - cache a reference to `Entities`
 - Duplicate the same checks as `Query::get` but with the cached references instead.

If this seems like a lot, I can probably do this in a followup PR.